apply/pipeline.go

```diff
@@ -50,114 +48,113 @@ func Pipeline(g *graph.Graph, id string, factory *render.Factory) executor.Pipel
+func (g *pipelineGen) maybeSkipApplication(resultI interface{}) (interface{}, error) {
+	if asResult, ok := resultI.(*Result); ok {
+		return asResult, nil
+	}
+	asPlan, ok := resultI.(resultWrapper)
+	if !ok {
+		return nil, fmt.Errorf("expected *Result or *resultWrapper but got type %T", resultI)
 	}
-	return monad.FMap(checkResult, resultI.(either.EitherM))
+	if !asPlan.Plan.Status.HasChanges() {
+		return &Result{
+			Ran:  false,
+			Task: asPlan.Plan.Task,
+			Plan: asPlan.Plan,
+			Err:  asPlan.Plan.Err,
+		}, nil
+	}
+	return asPlan, nil
 }
 
 // applyNode runs apply on the node, it takes an Either *apply.Result
 // *plan.Result and, if the input value is Left, returns it as a Right value,
 // otherwise it attempts to run apply on the *plan.Result.Task and returns an
 // appropriate Left or Right value.
-func (g *pipelineGen) applyNode(taski interface{}) monad.Monad {
-	taskE, ok := taski.(either.EitherM)
-	if !ok {
-		return either.LeftM(fmt.Errorf("expected either.EitherM but got %T", taski))
-	}
-	val, isRight := taskE.FromEither()
-	if !isRight {
-		return either.RightM(val)
+func (g *pipelineGen) applyNode(val interface{}) (interface{}, error) {
+	if asResult, ok := val.(*Result); ok {
+		return asResult, nil
 	}
 	twrapper, ok := val.(resultWrapper)
 	if !ok {
-		return either.LeftM(fmt.Errorf("apply expected a resultWrappert but got %T", val))
+		return nil, fmt.Errorf("apply expected a resultWrappert but got %T", val)
 	}
 	applyStatus, err := twrapper.Plan.Task.Apply()
 	if err != nil {
 		err = fmt.Errorf("error applying %s: %s", g.ID, err)
 	}
-	return either.RightM(&Result{
+	return &Result{
 		Ran:    true,
 		Status: applyStatus,
 		Task:   twrapper.Plan.Task,
 		Plan:   twrapper.Plan,
 		Err:    err,
-	})
+	}, nil
 }
 
 // maybeRunFinalCheck :: *Result -> Either error *Result; looks to see if the
 // current result ran, and if so it re-runs plan and sets PostCheck to the
 // resulting status.
-func (g *pipelineGen) maybeRunFinalCheck(resultI interface{}) monad.Monad {
+func (g *pipelineGen) maybeRunFinalCheck(resultI interface{}) (interface{}, error) {
 	result, ok := resultI.(*Result)
 	if !ok {
-		return either.LeftM(fmt.Errorf("expected *Result but got %T", resultI))
+		return nil, fmt.Errorf("expected *Result but got %T", resultI)
 	}
 	if !result.Ran {
-		return either.RightM(result)
+		return result, nil
 	}
 	task := result.Plan.Task
-	return plan.Pipeline(g.Graph, g.ID, g.RenderingPlant).
-		Exec(either.ReturnM(task)).
-		AndThen(func(planI interface{}) monad.Monad {
-			plan, ok := planI.(*plan.Result)
-			if !ok {
-				return either.LeftM(fmt.Errorf("expected *plan.Result but got %T", planI))
-			}
-			result.PostCheck = plan.Status
-			if plan.HasChanges() {
-				result.Err = fmt.Errorf("%s still has changes after apply", g.ID)
-			}
-			return either.RightM(result)
-		})
+	val, pipelineError := plan.Pipeline(g.Graph, g.ID, g.RenderingPlant).Exec(task)
+	if pipelineError != nil {
+		return nil, val.(error)
```

Comment:
there probably should have been a test to catch that, looks like it's a vestige of the monadic code, it should be `return nil, pipelineError`